### DISCUSSION TITLE
Contributing

### DISCUSSION THREAD

**Discussion by Hookyns** [Hookyns/tst-reflect#10] [General]
From @iDevelopThings, #9

> Hey, sorry I wanted to reach out, but couldn't find any contact info.
>
> I was working on a similar package with some other useful features, then I discovered yours, I really like what you've put together here
> 
> Is there anything you're against adding to this reflection wise and would you be open to me submitting prs with some of the features I'd love? :)
> 
> It doesn't make sense for me to release yet another package and such, it would make more sense for people with the same goal to work on a similar thing 😅
> 
> I was working on some reflection utilities but didn't get as far as you did, however I finished a system where interfaces have a type on runtime, which also have some reflection information - in a similar fashion to your "getType<Class/Interface>", it's possible to use "Interface" which the original declaration is replaced with an object and has some other things with it(still experimenting with the idea, but it allows for our interfaces to have a normal/available type on runtime, rather than vanishing from our code essentially) - I also did this for an IOC container implementation, similar to your goal here :D
> 
> If you'd like to discuss more, send me a dm on my twitter its on my profile... or we can discuss here, I don't mind

Hi, this is still a prototype, a lot of things can change but I'm open to contributions. I've not prepared any guide,.. but one important thing for me is keeping `Type` as close as possible to C#'s Type.

**Comment by iDevelopThings** [comment]
Yeah I understand that, and that's another reason I'd love to help, a project like this is a lot for one person, with some help, maybe one day it will be more than a prototype 😄 

And I'm cool with the C# Type, I'm not personally familiar with C# but I was reading the docs and it seems all good 

Is there anything you have on the todo list but cannot get to yet?

**Reply by iDevelopThings** [reply]
After reading more on the C# docs, I'd love to try come up with a solution for `.GetType()` on any kind of object. This would be fun :D

**Reply by Hookyns** [reply]
Well, if you want, there is https://github.com/Hookyns/ts-reflection/blob/31fec6c2f48482f3db4f073ba00b84e684d72201/transformer/src/getMethods.ts#L31-L40

Is there something you miss and want to implement?

C#'s Object.GetType() would be nice ~~but it requires extension of Object's prototype. I'm Okay with that but quite a lot of people don't like it, cuz it changes behavior of `for..in` on objects (x.hasOwnProperty() check required then; which is good practise anyway, just somebody doesn't like it.).~~
Well, if it will be handled by transformer and changed to getType() call, it is Okay.
But C#'s .GetType() is on object/variables and it returns real type of the value which is in the memory. We are not able to do that. We are able to get the type of the object/variable and return that type, but imagine this case:

```typescript
class Base { }
class Child extends Base {}

const variable: Base = new Child(); // type of the variable in type checking context is Base, not Child
```
or
```typescript
class Base { }
class Child extends Base {}
class Unrelated {}

const variable: Base = new Unrelated() as any;
```

You can test this:
```typescript
class Base { }
class Child extends Base {}

const variable: Base = new Child();

console.log(getType<typeof variable>()); // This is what we are capable of; typeof variable is Base
```

**Reply by Hookyns** [reply]
So wouldn't `variable.getType()` be misleading?

**Reply by iDevelopThings** [reply]
If the properties are defined like `Object.defineProperty(Base.prototype. "getType", {enumerable : false})` they don't appear in for loops etc

My idea was that for specific declarations, we can add `__getSymbol()` to the prototype which would output `Symbol.for()` with the unique id for this type, which would be handy for the Type runtime class also

Then `x.getType()` runs on the same basis, it can get it's symbol and manually call `getType(id)`, if I understand how it all works of-course

It could be missleading I agree, but if each declaration has a unique identifier, you'll theoretically be getting the correct type.... I think

I will take a look into the methods though for sure, no problem :)

---- 
On another side-note, are you able to share your editor formatting config? I was playing a little earlier and my autoformatting completely changes the structure, which I obviously don't want to be doing on your project 😅

**Reply by iDevelopThings** [reply]
Just as a quick example, it could be something like this... it's just a lot of code to be throwing in the transpiled version 

```typescript
declare global {
	interface Function {
		__getSymbol(): symbol;
		getType(): Type;
	}
	interface Object {
		__getSymbol(): symbol;
		getType(): Type;
	}
}

export class SomeClass {
	someMethod() {
		return "";
	}
}


Object.defineProperty(SomeClass, '__getSymbol', {
	enumerable : false,
	get(): any {
		return function () {
			return Symbol.for("485");
		}
	}
});
Object.defineProperty(SomeClass, 'getType', {
	enumerable : false,
	get(): any {
		return function () {
			return Type._getTypeMeta(Number(this.__getSymbol().description))
		}
	}
});

```

(this actually works with the id for the type :D)

**Reply by Hookyns** [reply]
> On another side-note, are you able to share your editor formatting config? I was playing a little earlier and my autoformatting completely changes the structure, which I obviously don't want to be doing on your project 😅

My formatting is affected by the C# formatting,.. .NET devs friendly, JavaScript devs unfriendly. 😅
I've just pushed updated .editorconfig but it is supported only by Jetbrains products (Webstorm, Rider etc.) 
ESLint is on the TODO list. Same as Github Actions cuz of PRs..

**Reply by iDevelopThings** [reply]
That's totally fine with me, I like the style of it. and perfect, I was hoping it was JetBrains based  :D

I will do some experimenting with my idea above, it seems there may be some flaws, jut realised that maybe I've mis-understood how javascript prototype works over all of the years 😅

**Reply by Hookyns** [reply]
There will be problem with Methods and similar things.. All of that can be used as Type. (I've implemented some of those things today :D)

```typescript
module mod {}

class A {
  doSomething();
}

let x: A["doSomething"] = function() {};

let m: typeof mod = {};
```

You want to generate those defineProperty, right? And make it available for everything? It means generate those defineProperty for every type in the source code; even those inside 3rd party packages...
Current implementation generate type information just for types passed to getType<T>(). Generating type information for everything is so OP for projects with thousands of dependencies.. 😮 

BTW: **What is your usage?** `SomeClass.getType()`? Well, the C#'s usage is `var x = new SomeClass(); x.GetType();`. It should be called on the instance of an Object.

So it would be:
```typescript
Object.defineProperty(Object.prototype, 'getType', {
	enumerable : false,
	get(): any {
		return function () {
			return Type._getTypeMeta(Number(this.__getSymbol().description))
		}
	}
});
```
And generate only those __getSymbol().

**Reply by Hookyns** [reply]
IMHO, it will be much easier to handle all `.getType()` method calls an process that the same way as `getType<T>()` function calls.

And just add global declaration into the tst-reflect package:
> global.d.ts
```typescript
interface Object
{
	getType: (() => import("tst-reflect").Type) & { __tst_reflect__: true }; // that object with __tst_reflect__ is for identification of our `.getType()`, cuz somebody can override that with own method.
}
```
and 
> reflect.ts
```typescript
Object.defineProperty(Object.prototype, "getType", {
	enumerable: false,
	get(): any
	{
		return function () {
			return undefined;
		};
	}
});
```

And final usage: 
```typescript
class A {}

const a = new A();

a.getType(); // handled by transformer and replaced by tst_reflect.getType(12345)
```


PS: Maybe I just missed your point,.. write me some example, what to change in runtime, transformer and what will be the usage.

**Comment by iDevelopThings** [comment]
Yeah this is a good point, it's not great to generate it for everything 😅 

Only reason I thought of doing this is because it was a nice thing to have access too, but also came with the `__getSymbol()` which would be extremely useful for resolving other types on the fly
From what I could understand when playing with it, there's no way for the Type class to for example resolve some other type on the fly, without it being passed in right? 
At-least `__getSymbol()` would give some way to access this, but I also think this wouldn't technically be possible anyway since as I've read somewhere here on the repo, there's no "global" store for the type data it's essentially just what is picked up on the fly and replaced in file?

I wont worry too much about these things for now any-way, will help get some of the other parts of the package that are probably more critical 

Now that you mention it though, I do also prefer this syntax a lot :D

```typescript
class A {}

const a = new A();

a.getType(); // handled by transformer and replaced by tst_reflect.getType(12345)
```

**Reply by Hookyns** [reply]
> I've read somewhere here on the repo, there's no "global" store for the type data it's essentially just what is picked up on the fly and replaced in file?

Yes and no. 

Base(/first implemented) mode is per-file "cache". Each type's info is generated once per file, no matter how many getType<T> calls are there. Because in TS you don't know which file is "root" (executed/loaded first), so you don't know where to place all the generated metadata.

## Example of the base per-file mode:
> tsconfig.json
```json5
	"reflection": {
		"metadata": false, // disable second "metadata" mode, which is under development but it is default mode now, cuz it is better
		"debugMode": false
	}
```

> index.ts
```typescript
import {
	getType,
	Type,
	TypeKind
} from "tst-reflect";

abstract class Animal
{
	abstract get kind(): string;
	abstract get name(): string;

	protected constructor(protected readonly _kind: string)
	{
	}

	abstract makeSound();
}

const typeOfAnimal: Type = getType<Animal>();

console.log("Kind of the Type: ", TypeKind[typeOfAnimal.kind], " Name of the Type:", typeOfAnimal.name, " Full name of the Type:", typeOfAnimal.fullName, " Details:", typeOfAnimal);

console.log(getType<Animal>().name);
console.log(getType<Animal>().fullName);
console.log(getType<Animal>().baseType);
```

Generated JS:
> index.js
```javascript
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tst_reflect_1 = require("tst-reflect");
tst_reflect_1.getType({ n: "Animal", fn: "..\\index.ts:Animal", props: [{ n: "kind", t: tst_reflect_1.getType({ n: "string", k: 2 }) }, { n: "name", t: tst_reflect_1.getType({ n: "string", k: 2 }) }, { n: "_kind", t: tst_reflect_1.getType({ n: "string", k: 2 }) }], ctors: [{ params: [{ n: "_kind", t: tst_reflect_1.getType({ n: "string", k: 2 }) }] }], k: 1, ctor: () => Animal }, 23251);
class Animal {
    constructor(_kind) {
        this._kind = _kind;
    }
}
const typeOfAnimal = tst_reflect_1.getType(23251);
console.log("Kind of the Type: ", tst_reflect_1.TypeKind[typeOfAnimal.kind], " Name of the Type:", typeOfAnimal.name, " Full name of the Type:", typeOfAnimal.fullName, " Details:", typeOfAnimal);
console.log(tst_reflect_1.getType(23251).name);
console.log(tst_reflect_1.getType(23251).fullName);
console.log(tst_reflect_1.getType(23251).baseType);
//# sourceMappingURL=index.js.map
```

Transformer goes through source code and looks for `getType<T>()` calls. If there is an `getType` call, it generates object with type details, puts it at the top of current file and replaces that `getType<T>()`call by runtime call of `getType(id)` with type's id as argument.
Each call for same type is replaced by the same `getType(id)` call.

## Second "metadata" mode (work in progress, buggy)
Enabled by default as I wrote in first section.
That mode generates all the types' metadata into one metadata file which must be imported by user/programmer in root module.
All the types are in that metadata library, but only those requested by `getType<T>()` calls.

_I wanted to implement config option to enable "full" metadata library, it means generate metadata for all of the types in the project (including packages) but cuz of performance, it is not a good idea.._

**Reply by iDevelopThings** [reply]
Okay I see now, I think I tried the second option at some point, which didn't work for me, but now I think back... I don't think I imported the generated file 😅 

`I wanted to implement config option to enable "full" metadata library, it means generate metadata for all of the types in the project (including packages) but cuz of performance, it is not a good idea..`

Yeah... I can see that, it might be a good idea to be able to configure some paths which will be generated for, regardless of it, kind of like typescripts tsconfig "include" option

If for example, with your container, you made a package, but had some logic depending on reflection there, it would cause some problems and not work, or in a mono-repo style project

I'd be open to put some more work into this metafile style generator once I'm done with method reflection :)

**Reply by Hookyns** [reply]
The biggest issue in metadata lib mode are imports of classes for `ctor`. Check https://github.com/Hookyns/ts-reflection/blob/c2242921fec1d226f8a3482a8112612be10e9bff/transformer/src/getTypeDescription.ts#L299
Excluding this main issue, it should work quite well.
